provider/etcd/etcd_client.go

```diff
@@ -0,0 +1,337 @@
+// Copyright 2016 Prometheus Team
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+
+How the Etcd Alerts Storage Provider Works
+
+Initialization (Synchronous)
+
+* On startup, Alertmanager tries to connect to Etcd.  If the connection attempt
+  is unsuccessful after a timeout, then Alertmanager fails hard because it is
+  probably a configuration error.  A supervisor such as Kuberntes should
+  restart it.
+
+* Once Alertmanager successfully connects to Etcd, it proceeds with a "best
+  effort" to write alerts to Etcd as well receive alert updates Etcd via a
+  watch client.  If Etcd disappears for any reason (e.g. network partition,
+  quorum loss, etc.), Alertmanager will continue running independently to
+  maintain availability.  Once connectivity to Etcd is restored, Alertmanager
+  will resume writing alerts to Etcd and receiving alert updates from Etcd.
+
+* Alertmanager starts an Etcd watch client to receive alert updates from Etcd.
+  As other Alertmanager in the Alertmanager cluster write to Etcd, all of the
+  Alertmanagers receives those updates.
+
+* For the last part of initialization, Etcd will proceed to read all alert
+  objects from Etcd.
+
+Ongoing Operation (Async)
+
+* Alertmanager can receive alerts from both API requests AND the Etcd watch
+  subscription.  Alerts from both of these sources are sent to the etcd alert
+  provider via Put().  The alert is put into Etcd only when there is a
+  difference between the Etcd alert and memory.
+
+* The Etcd clientv3 library will automatically reconnect if the Etcd server
+  goes down or up, or upon resuming network connectivity if it had been lost.
+
+* The long running watch subscription will reconnect if the watch channel is
+  closed.
+
+*/
+
+package etcd
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"reflect"
+	"sync"
+	"time"
+
+	"go.etcd.io/etcd/clientv3"
+	"google.golang.org/grpc"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/alertmanager/types"
+)
+
+var (
```

Comment:
SHOULD: Use const for error, but perhaps that is not the pattern in this repo.

ref: https://dave.cheney.net/2016/04/07/constant-errors

Reply:
Leaving it, since that's how it was other places in AM.